### PR TITLE
Annotations reference workflow

### PR DESCRIPTION
**Pull request by wild-endeavor**

### CODE REVIEW

File: flytekit/annotated/workflow.py [hunk 1]
```diff
@@ -289,11 +291,57 @@ def get_registerable_entity(self) -> _SdkWorkflow:
         return self._registerable_entity
 
 
-def workflow(_workflow_function=None):
+class ReferenceWorkflow(Workflow):
```
Comment (kumare3) [inline flytekit/annotated/workflow.py:294]: should we make it such that a ReferenceWorkflow can be created inline, without needing a function. Something like this
```python
w = Workflow(reference=WorkflowReference(...), inputs=kwtypes(), outputs=kwtypes())
```

Comment (kumare3) [reply flytekit/annotated/workflow.py:294]: I know you made the task also, to use an empty function, but if we factor it so, then we can implement both methods?

File: tests/flytekit/unit/annotated/test_type_hints.py [hunk 2]
```diff
@@ -1057,3 +1083,33 @@ def my_wf(a: int) -> str:
 
     with pytest.raises(TypeError):
         my_wf(a=5)
+
+
+def test_fff():
```
Comment (kumare3) [inline tests/flytekit/unit/annotated/test_type_hints.py:1088]: can we name it better?

File: flytekit/annotated/testing.py [hunk 3]
```diff
@@ -35,3 +37,28 @@ def _log(*args, **kwargs):
...
+    if not isinstance(target, PythonTask) and not isinstance(target, Workflow):
+        raise Exception("Can only use mocks on tasks/workflows declared in Python.")
+
+    def wrapper(test_fn):
+        def new_test(*args, **kwargs):
+            logger.warning(f"Invoking mock method for task: '{target.name}'")
+            m = MagicMock()
+            if isinstance(target, PythonTask):
+                saved = target.execute
+                target.execute = m
```
Comment (kumare3) [inline flytekit/annotated/testing.py:52]: I made a dummy method that logs that the mock has been invoked,  do you want to add that here to? Helps in debugging IMO

File: flytekit/annotated/testing.py [hunk 4]
```diff
@@ -35,3 +37,28 @@ def _log(*args, **kwargs):
     t.execute = _log
     yield m
     t.execute = _captured_fn
+
+
+def patch(target: Union[PythonTask, Workflow]):
+    if not isinstance(target, PythonTask) and not isinstance(target, Workflow):
```
Comment (kumare3) [inline flytekit/annotated/testing.py:43]: like this type-check. Can you add it to `task_mock` as well?

**Comment by kumare3** [comment]
couple comments, then lgtm

File: flytekit/annotated/testing.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/annotated/testing.py:52]: copied the warning above

File: flytekit/annotated/workflow.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/annotated/workflow.py:293]: How about just a `ReferenceEntity` but keep the other ones as is?

File: flytekit/annotated/testing.py (hunk 4 above)
Comment (wild-endeavor) [reply flytekit/annotated/testing.py:43]: sure.

File: flytekit/annotated/testing.py (hunk 3 above)
Comment (kumare3) [reply flytekit/annotated/testing.py:56]: where is the copy?

Comment (wild-endeavor) [reply flytekit/annotated/testing.py:56]: line 52

File: flytekit/annotated/testing.py [hunk 5]
```diff
@@ -35,3 +41,29 @@ def _log(*args, **kwargs):
...
+        raise Exception("Can only use mocks on tasks/workflows declared in Python.")
+
+    def wrapper(test_fn):
+        def new_test(*args, **kwargs):
+            logger.warning(f"Invoking mock method for task: '{target.name}'")
```
Comment (wild-endeavor) [inline flytekit/annotated/testing.py:52]:
```suggestion
            logger.warning(f"Invoking mock method for target: '{target.name}'")
```

**Pull request merged**